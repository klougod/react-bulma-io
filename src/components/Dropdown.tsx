import React, {
  useState,
  useEffect,
  useRef,
  Children,
  ReactElement
} from 'react'
import PropTypes from 'prop-types'
import { faAngleDown } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

export const Dropdown = ({
  className,
  btnClassName,
  btnTitle,
  customBtn,
  children
}: any) => {
  const [toggle, setToggle] = useState(false)
  const wrapperNode: any = useRef(null)

  useEffect(() => {
    document.addEventListener('mousedown', handleClick)
    return () => {
      document.removeEventListener('mousedown', handleClick)
    }
  }, [wrapperNode])

  const handleClick = (e: MouseEvent) => {
    if (wrapperNode) {
      if (!wrapperNode.current.contains(e.target as Node)) {
        setToggle(false)
      }
    }
  }

  const Button = () => (
    <button
      className={`button ${btnClassName}`}
      aria-haspopup='true'
      aria-controls='dropdown-menu'
    >
      <span>{btnTitle}</span>
      <span className='icon is-small'>
        <FontAwesomeIcon icon={faAngleDown} />
      </span>
    </button>
  )

  return (
    <div
      ref={wrapperNode}
      className={`dropdown ${className} ${toggle && 'is-active'}`}
      onClick={() => setToggle(!toggle)}
    >
      <div className='dropdown-trigger'>
        {customBtn ? customBtn : <Button />}
      </div>
      <div className='dropdown-menu' id='dropdown-menu' role='menu'>
        <div className='dropdown-content'>
          {Children.toArray(children).map((item: ReactElement, index) =>
            item.type !== 'hr'
              ? React.cloneElement(item, {
                  key: index,
                  className: `${
                    item.props.className ? item.props.className : ''
                  }dropdown-item`
                })
              : React.cloneElement(item, {
                  key: index,
                  className: 'dropdown-divider'
                })
          )}
        </div>
      </div>
    </div>
  )
}

Dropdown.defaultProps = {
  className: '',
  btnClassName: '',
  btnTitle: ''
}

Dropdown.propTypes = {
  className: PropTypes.string,
  btnClassName: PropTypes.string,
  btnTitle: PropTypes.string,
  children: PropTypes.node.isRequired
}

export default Dropdown
