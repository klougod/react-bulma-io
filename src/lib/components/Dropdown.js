import React, { useState, useEffect, useRef, Children } from 'react'
import PropTypes from 'prop-types'
import { faAngleDown } from "@fortawesome/free-solid-svg-icons"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"

export const Dropdown = ({ className, btnClassName, btnTitle, customBtn, children }) => {
  const [toggle, setToggle] = useState(false)
  const wrapperNode = useRef(null)

  useEffect(() => {
    document.addEventListener('mousedown', handleClick)
    return () => {
      document.removeEventListener('mousedown', handleClick)
    };
  }, [wrapperNode])

  const handleClick = e => {
    if (wrapperNode.current.contains(e.target)) {
      return
    }
    setToggle(false)
  }

  const Button = () => (
    <button className={`button ${btnClassName}`} aria-haspopup="true" aria-controls="dropdown-menu">
      <span >{btnTitle}</span>
      <span className="icon is-small">
        <FontAwesomeIcon icon={faAngleDown} />
      </span>
    </button>
  )

  return (
    <div ref={wrapperNode} className={`dropdown ${className} ${toggle && "is-active"}`} onClick={() => setToggle(!toggle)}>
      <div className="dropdown-trigger">
        {customBtn ? customBtn : <Button />}
      </div>
      <div className="dropdown-menu" id="dropdown-menu" role="menu">
        <div className="dropdown-content">
          {Children.toArray(children).map((item, index) =>
            item.type !== "hr" ?
            React.cloneElement(item, {
              key: index,
              className: `${item.props.className ? item.props.className : ''}dropdown-item`
            }) :
            React.cloneElement(item, {
              key: index,
              className: 'dropdown-divider'
            })
          )}
        </div>
      </div>
    </div>
  )
}

Dropdown.defaultProps = {
  className: '',
  btnClassName: '',
  btnTitle: ''
}

Dropdown.propTypes = {
  className: PropTypes.string,
  btnClassName: PropTypes.string,
  btnTitle: PropTypes.string,
  children: PropTypes.node.isRequired
}

export default Dropdown