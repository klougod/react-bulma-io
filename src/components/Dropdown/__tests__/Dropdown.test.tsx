import React from 'react'
import Enzyme from 'enzyme'

import Dropdown from '../'

describe('Dropdown component test', function () {
  it('should render correctly without props', function () {
    const dropdownMock = Enzyme.shallow(<Dropdown />)
    expect(dropdownMock).toMatchSnapshot()
  })

  it('should render correctly with props', function () {
    const dropdownMockProps = {
      className: 'dropdownClass',
      btnClassName: 'dropdownBtnClass',
      btnTitle: 'dropdownBtnTitle'
    }
    const dropdownMock = Enzyme.shallow(<Dropdown {...dropdownMockProps} />)
    expect(dropdownMock).toMatchSnapshot()
  })

  it('should render correctly with custom button', function () {
    const dropdownMockProps = {
      customBtn: <button />
    }
    const dropdownMock = Enzyme.shallow(<Dropdown {...dropdownMockProps} />)
    expect(dropdownMock).toMatchSnapshot()
  })

  it('should call onclick function prop if dropdown is clicked', function () {
    const dropdownMockProps = { onClick: jest.fn() }
    const dropdownMock = Enzyme.shallow(<Dropdown {...dropdownMockProps} />)
    dropdownMock
      .find('div')
      .at(0)
      .simulate('click', { target: {} })
    expect(dropdownMock).toMatchSnapshot()
  })
})
