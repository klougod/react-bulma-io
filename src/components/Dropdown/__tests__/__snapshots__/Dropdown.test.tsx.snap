// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component test should render without crash with props 1`] = `
<div
  className="dropdown dropdownClass false"
  onClick={[Function]}
>
  <div
    className="dropdown-trigger"
  >
    <Button />
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    />
  </div>
</div>
`;

exports[`Dropdown component test should render without crash without props 1`] = `
<div
  className="dropdown  false"
  onClick={[Function]}
>
  <div
    className="dropdown-trigger"
  >
    <Button />
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    />
  </div>
</div>
`;
